Add HitShape trait

reaperrr opened the pull request:
...move Health.Shape there, add `ITargetablePositions`, and remove `ITargetablePositions` from `Building`.

This is the easier, uncontroversial part of #13196 and should *not* cause any regressions on bleed.

Closes #7717.
Closes #10349.

reaperrr commented:
I've come up with an idea for asymmetric footprint hit-shapes, based on a comment by @obrakmann on #10383.

Basically, what I've done on a local branch is moving `IHitShape` from `Health` to `TargetableLocations`.

This brings a number of advantages:
- you can have multiple shapes per actor, without the need to refactor the shapes themselves
- you can enable/disable them at will via conditions
- you can have actors that block projectiles without the requirement for a health trait

Example, using the RA refinery:
![ra-ref](https://cloud.githubusercontent.com/assets/2857877/26201732/a89eb302-3bd4-11e7-83ea-74922513be21.png)

pchote commented:
What about actors that can take aoe damage but is not itself targetable?

reaperrr commented:
I guess you're right, although I can't come up with any examples (unless bridges fall into that category already?).
I still think we should untie shapes from `Health` though, as the above advantages would still apply.

Maybe we should move it to a dedicated `HitShape` trait, afterall? That way, we and/or 3rd-party modders could make other stuff like `Shield` traits etc. interact with shapes, while avoiding clashes with `Health`.


Anyway, I think we can or even should leave this PR dedicated to its original purpose and deal with any shape trait refactoring in its own PR, but in my opinion multiple rectangle shapes are still a better solution for assymetric structures in the legacy mods than resorting to skewing shapes (that's still a desired feature on its own, I'm just not convinced we should use it for the legacy mods).

GraionDilach commented:
>What about actors that can take aoe damage but is not itself targetable?

IMO those should use a custom strictly-warheads targettype to be damaged. I plan to have such a system where actors can gain such untargetability from a weapon jammer actions (following examples I've seen in Rise of the Reds, like the Shtora of the Golem or the Anvil Bot).

reaperrr committed:
Move Shape from Health to new HitShape trait

Renamed Shape to Type

reaperrr committed:
Upgrade rule for Health.Shape to HitShape.Type

reaperrr commented:
> What about actors that can take aoe damage but is not itself targetable?

Actually, now that I've thought about it some more, I think that wouldn't be a problem.

All that `TargetableLocations` does by default, without customization, is provide a single targetable position at the actors' `CenterPosition`. That's what the `Target` code falls back to on all actors, anyway. `TargetableLocations` does not really make the actor targetable afterall, that's what `Targetable` is for.

Basically, moving HitShapes to `TargetableLocations` and then adding that trait to the actors in your example should not have any impact, let alone cause any regressions, because the `CenterPosition` of any actor is considered a 'targetable position' on bleed already.


*However*, the actual problems I see are that 
a) having the shape reside on `TargetableLocations` might be a little unintuitive to some modders, and 
b) making shapes disableable via conditions will likely be tricky due to the way they are loaded (out of scope for now, at least), so having them on a conditional trait when they're not disableable would be misleading.

Keeping the shape a separate trait is probably the 'safer' and more future-proof approach.

reaperrr committed:
Lint check and yaml enforcement for HitShape

Now that Health no longer provides a HitShape, actors with Health need at
least one HitShape trait.

pchote commented:
>having the shape reside on `TargetableLocations` might be a little unintuitive to some modders

This is just a naming issue, which goes away if we e.g. rename `TargetableLocations` to `HitShape` or `HitBox` `Health.Shape` to `HitShape.Type`.

> making shapes disableable via conditions will likely be tricky due to the way they are loaded

I don't see the problem here - we load all of them exactly once, and then pick between them at runtime?

reaperrr commented:
Yeah, the former occured to me yesterday as well, and the latter was actually just a brain snafu + mistake on my local branch (I forgot that `Exts.IsTraitEnabled` doesn't work on TraitInfos, so my shapes always stayed enabled).

I just can't log into gh from work so I didn't have the chance to update my comment before yours.

About the names, I think `HitShape` is better than `HitBox` (circles and capsules aren't boxes, but certainly shapes), I'm fine with `Shape` -> `Type`.

reaperrr committed:
Add TargetableOffsets to HitShape

And remove ITargetablePositions from Building.
Also, added UseFootprintOffsets to replicate the old Building behavior for easier setup of TargetablePositions for buildings.

reaperrr commented:
I'll update this after #13340 has been merged.

reaperrr commented:
Updated, though it still depends on #13340.

atlimit8 commented:
#13340 has been merged.

reaperrr committed:
Fix inconsistent rules in preparation of HitShape refactor

- TD C17 and A10 are not targetable and therefore technically invulnerable, so we should remove Health and Armor
- D2k ^carryall.colorpicker and TS ^mmch.colorpicker shouldn't have Health
- TS crash sites are supposed to be indestructible map deco as per original (they had "Immune=yes")

reaperrr commented:
Rebased.

pchote commented:
I hit a crash when starting a game on "Dual Cold Front" map in RA:
```
Exception of type `System.InvalidOperationException`: Actor does not have trait of type `OpenRA.Mods.Common.Traits.BodyOrientation`
```
Not sure which actor is causing it.

pchote commented:
> - TS crash sites are supposed to be indestructible map deco as per original (they had "Immune=yes")

I remember failing missions where you had to protect the crash sites from being destroyed...?  In any case, it will be a long time before we have to worry about these missions so I see no reason not to remove the health here for now.

pchote commented on OpenRA.Mods.Common/Projectiles/Bullet.cs:
```diff
@@ -302,12 +303,12 @@ bool AnyValidTargetsInRadius(World world, WPos pos, WDist radius, Actor firedBy,
 
-				var healthInfo = victim.Info.TraitInfoOrDefault<HealthInfo>();
-				if (healthInfo == null)
+				var activeShapes = victim.TraitsImplementing<HitShape>().Where(Exts.IsTraitEnabled).ToArray();
```
This can be made more efficient by removing the `ToArray` and the `if (!activeShapes.Any())`.  The `activeShapes.Any(i => ...)` check can then return before having to fully evaluate the enumerable.

pchote commented on OpenRA.Mods.Common/Traits/CombatDebugOverlay.cs:
```diff
@@ -75,6 +74,10 @@ void IRenderAboveWorld.RenderAboveWorld(Actor self, WorldRenderer wr)
 				TargetLineRenderable.DrawTargetMarker(wr, hc, hb);
 			}
 
+			var activeShapes = shapes.Where(Exts.IsTraitEnabled);
```
You might as well move this inline, it's short and you're only using it once.

reaperrr replied:
I actually use if for `var positions` in line 80, too.

pchote commented on OpenRA.Mods.Common/Traits/HitShape.cs:
```diff
@@ -0,0 +1,60 @@
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Shape of actor for damage calculations.")]
```
"Shape of actor for targeting and damage calculations."?

pchote commented on OpenRA.Mods.Common/Traits/HitShape.cs:
```diff
@@ -0,0 +1,60 @@
+
+			if (!string.IsNullOrEmpty(shape))
+			{
+				ret = Game.CreateObject<IHitShape>(shape + "Shape");
```
This probably wants to live inside the try so we can give a sensible exception if the shape type doesn't exist.

pchote commented on OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs:
```diff
@@ -88,12 +89,12 @@ public bool GetDirectHit(World world, CPos cell, WPos pos, Actor firedBy, bool c
 
-				var healthInfo = victim.Info.TraitInfoOrDefault<HealthInfo>();
-				if (healthInfo == null)
+				var activeShapes = victim.TraitsImplementing<HitShape>().Where(Exts.IsTraitEnabled).ToArray();
```
Likewise here: drop the `ToArray` and `if (!Any)`.

pchote commented on OpenRA.Mods.Common/Warheads/SpreadDamageWarhead.cs:
```diff
@@ -66,11 +67,17 @@ public override void DoImpact(WPos pos, Actor firedBy, IEnumerable<int> damageMo
 
-				var distance = healthInfo.Shape.DistanceFromEdge(pos, victim);
+				// Cannot be damaged without an active HitShape
+				var activeShapes = victim.TraitsImplementing<HitShape>().Where(Exts.IsTraitEnabled).ToArray();
```
For transient collections that you only enumerate once `ToArray` actually hurts performance (you need to enumerate it twice to create the array, and then a third time to do the thing you want to do).

pchote commented on OpenRA.Mods.Common/WorldExtensions.cs:
```diff
@@ -45,9 +46,9 @@ public static IEnumerable<Actor> FindActorsOnLine(this World world, WPos lineSta
-				var healthInfo = currActor.Info.TraitInfoOrDefault<HealthInfo>();
-				if (healthInfo != null)
-					actorWidth = healthInfo.Shape.OuterRadius.Length;
+				var shapeInfos = currActor.Info.TraitInfos<HitShapeInfo>().Where(Exts.IsTraitEnabled);
```
`Where(Exts.IsTraitEnabled)` doesn't make sense on TraitInfos.  You will need to enumerate `TraitsImplementing<HitShape>().Where(Exts.IsTraitEnabled)` and then use `.Info.Shape`

reaperrr replied:
Yeah, I generally noticed a while ago but obviously forgot to update this.

pchote commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -679,6 +679,31 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
+						{
+							var hitShape = new MiniYamlNode("HitShape", "");
+							node.Value.Nodes.Add(hitShape);
+							Console.WriteLine("'Shape' was moved from Health to HitShape and renamed to 'Type'.");
```
Do modders need to do any manual work for this?  If not, best to keep it in the changelog instead of logging it tens of times during the upgrade.

pchote commented on OpenRA.Mods.Common/Traits/HitShape.cs:
```diff
@@ -19,6 +19,12 @@ namespace OpenRA.Mods.Common.Traits
 	[Desc("Shape of actor for damage calculations.")]
 	public class HitShapeInfo : ConditionalTraitInfo
```
You `self.Trait<BodyOrientation>()`, so this needs a `Requires<BodyOrientationInfo>`.

reaperrr replied:
Now that I think about it, is it actually correct to require `BodyOrientation` here? I mean, bridges etc. don't need a `BodyOrientation` otherwise, I could instead adapt the `TargetablePositions` code below to only rotate the offset if the actor has a `BodyOrientation`.

pchote replied:
`BodyOrientation` is used to translate between local and world coordinates, which is important even for stationary actors because that conversion might need camera angle or perspective fudges.

reaperrr replied:
Ok, in that case I'll add a BO check to the lint rule as well.

pchote replied:
The `Requires` here already does that.

pchote commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -704,6 +704,18 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
+					var building = node.Value.Nodes.FirstOrDefault(n => n.Key == "Building");
+					if (building != null)
+					{
+						Console.WriteLine("The 'Building' trait no longer provides per-cell targetable positions.");
```
IMO it would be better to unconditionally add a `HitShape` node with `UseFootprintOffsets: true`.  It will then merge in with any other `HitShapes` and hopefully just work.

pchote commented on OpenRA.Mods.Common/Traits/HitShape.cs:
```diff
@@ -0,0 +1,93 @@
+		{
+			IHitShape ret;
+
+			var shapeNode = yaml.Nodes.Find(n => n.Key == "Type");
```
`.FirstOrDefault` please.

pchote reviewed (dismissed):
Looks mostly good.  Here's a few code issues that I noticed.

reaperrr commented:
> I remember failing missions where you had to protect the crash sites from being destroyed...? In any case, it will be a long time before we have to worry about these missions so I see no reason not to remove the health here for now.

Those `CACRSHxx` are just small, 1-cell crashed helis etc., you're mixing them up with the Scrin Crash site.

Those small map deco crash sites never had any mission relevance and were always indestructible, unselectable, untargetable etc.

reaperrr commented:
> you're mixing them up with the Scrin Crash site.

I just remembered, you're most likely actually talking about missions where you had to defend GDI Kodiak (GAKODK) or Nod Montauk (NAMNT) during ion storms. Those don't have anything to do with CACRSHxx either, though, the latter were really only used as map decoration.

reaperrr commented:
Updated, adressed everything that was pointed out.

atlimit8 commented on OpenRA.Mods.Common/Traits/HitShape.cs:
```diff
@@ -0,0 +1,92 @@
+			if (IsTraitDisabled)
+				yield break;
+
+			if (Info.UseFootprintOffsets && occupy != null)
```
`UseFootprintOffsets` should be `UseOccupiedCellsOffsets` since it uses occupied cells rather than a footprint per se.

reaperrr commented:
Updated.

pchote reviewed (approved):
Code looks good now, and I didn't notice any issues during quick playtests in TS and TD. :+1:

pchote commented:
This is one of our highest priority open PRs, so can another @OpenRA/engine-hackers please review ASAP?

cjshmyr reviewed (approved):
Looks good here and in game (TS was untested because of the armament NRE).

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/717bd77bf756484feb99d9f0f816c4f905195df0...4fe0b3d863d8a7830e8ca0753b161b9700bba53c
